README.rst
Dash Electrum - Lightweight Dashpay client
=====================================

::

  Licence: MIT Licence
  Author: Thomas Voegtlin
  Language: Python (>= 3.6)
  Homepage: https://electrum.dash.org/


.. image:: https://travis-ci.org/ktechmidas/electrum-dash.svg?branch=master
    :target: https://travis-ci.org/ktechmidas/electrum-dash
    :alt: Build Status





Getting started
===============


Use PPA setup
-------------

On Ubuntu/Linux Mint you can try to install Dash Electrum with next commands::

    sudo add-apt-repository ppa:Bertrand256/dash-electrum
    sudo apt-get update
    sudo apt-get install dash-electrum


Use source distribution
-----------------------

Dash Electrum is a pure python application. If you want to use the
Qt interface, install the Qt dependencies::

    sudo apt-get install python3-pyqt5

If you downloaded the official package (tar.gz), you can run
Dash Electrum from its root directory without installing it on your
system; all the python dependencies are included in the 'packages'
directory (except x11-hash).

To install x11-hash dependency in the 'packages' dir run once::

    python3 -m pip install -t packages x11-hash

To install precise tested versions of HW libs (trezor, ledeger, etc) run once::

    python3 -m pip install -t packages -r contrib/deterministic-build/requirements-hw.txt

To install precise tested version of pyqt5 run once::

    python3 -m pip install -t packages -r contrib/deterministic-build/requirements-binaries.txt

To run Dash Electrum from its root directory, just do::

    ./electrum-dash

You can also install Dash Electrum on your system, by running this command::

    sudo apt-get install python3-setuptools
    python3 -m pip install .[fast]

This will download and install the Python dependencies used by
Dash Electrum instead of using the 'packages' directory.
The 'fast' extra contains some optional dependencies that we think
are often useful but they are not strictly needed.

If you cloned the git repository, you need to compile extra files
before you can run Dash Electrum. Read the next section, "Development
Version".


Using Tor proxy
===============

Starting from Dash Electrum release 3.2.3.1 automatic Tor Proxy
detection and use on wallet startup is added to
`Network <docs/tor/tor-proxy-on-startup.md>`_ preferences.

To use Tor Proxy on Ubuntu set it up with::

    sudo apt-get install tor
    sudo service tor start

Other platforms setup is described at `docs/tor.md <docs/tor.md>`_

Development version
===================

Check out the code from GitHub::

    git clone https://github.com/MirzoyanGevorg/electrum-dash.git
    cd electrum-dash

Run install (this should install dependencies)::

    python3 -m pip install .[fast]


Compile the protobuf description file::

    sudo apt-get install protobuf-compiler
    protoc --proto_path=electrum_dash --python_out=electrum_dash electrum_dash/paymentrequest.proto

Create translations (optional)::

    sudo apt-get install python-requests gettext
    ./contrib/make_locale


RELEASE-NOTES
# Release 4.1.5.1

 * trezor plugin: allow skip session clear on wallet close #292
 * storage: more write attempts, write attempts GUI #293, #296
 * simple_config.py: fix FEERATE_STATIC_VALUES not to be less 1000 #295
 * Kivy: make the PS shield button simply start mixing #310
 * PrivateSend: optionally allow ps coinchooser to spend high fees #311
 * Kivy: allow switch to digits only password dialog #313
 * Kivy: add "Quick Scan", "Quick Receive" to main screen #314
 * Kivy: fix PrivateSend mixing progress refresh #315
 * PrivateSend: add PS button icon for stopping/starting states #316
 * Enable auto crash reports (copy to clipboard mode) #319
 * Add PayServer HTML/JS to electrum_dash/www #321
 * Qt: add warning on possible lags on HW devices #322
 * PrivateSend: fix freeze coins/addresses functionality #324
 * Kivy: unite Dash network static peers setttings #325
 * Kivy: use popups for Tx sending/error messages #326
 * Network: allow blacklist electrum servers #327


# Release 4.1.5.0

 * Kivy/Android: fix paying bip70 invoices (regression) (90579ccf)
 * fix: payment requests not saved if process is killed (6a049d99)
 * Qt: allow restoring from SLIP39 seeds (#6917)
 * Qt: rework QR code scanning on Windows and macOS (#7365)
 * Qt: support smaller window sizes, decrease minimums (#7385)
 * GUIs: add "funded or unused" filter option to Addresses tab (#5823)
 * CLI/RPC: fix "close_wallet" command (#7348)
 * Android: the binaries (APKs) prepared to be reproducible
 * AppImage: fix some startup issues by including libxcb deps (#7198)
 * imported wallets: when enabling the "Use change addresses" option,
   change will now be sent to a random unused imported address. (#7330)
   As before, by default, change is sent back to the "from address".
 * seed generation: make sure newly created electrum seeds don't have
   correct bip39 checksum by chance (#6001)
 * other minor fixes


# Release 4.1.2.2

 * wallet/kivy: fix feerate calc (float Decimal conversion) #258
 * PS: fix processing of new_denoms/collateral wfls #261
 * PS: do not check enough small denoms for ABS calc method #262
 * network: do not stop dash_net/mn_list if not full_shutdown #263
 * PS: add gathering mixing statistics #266
 * Qt fix: remove BlockingWaitingDialog use, other qt fixes #268
 * protx/qt: allow add MN with IP/BLS pubk found in registered MNs #269
 * protx_list: MNList.reset: notify gui, clear protx_info data #270
 * dash_net: rm staled [test.]dnsseed.masternode.io #271
 * Enhance PS gui logging about session timeouts #272
 * Kivy: fix start mixing with protected method #275


# Release 4.1.2.1

 * base_wizard.py, keystore.py: add purpose48 multisig #244
 * dash_ps: enhance new denoms/collateral patterns #246
 * commands load_wallet: add set current for commands option #247
 * dash_ps: fix _calc_total_need_val #248
 * qt dip3: make wizards modeless #251
 * Fix kivy: show tx type again #252
 * base_wizard.py: fix choose_keystore (create seed) #253
 * kivy: add network data loading progress bar, notify
   when server disconnected on excessive resource usage (limits) #254
 * Renew build dependencies for macOS/Win, other fixes #255
 * Win builds: custom PyInstaller bootloaders to minimize
   AV false-positives #256


# Release 4.1.2.0

 * The wallet creation wizard no longer asks for a seed type.
 * Paid invoices are removed from the lists. Invoice details can still
   be accessed from the transaction history.
 * Password unification (Android only): When the Android app is
   started, the entered password is checked against all wallets in
   the directory. If the test passes:
    - all wallets are encrypted
    - new wallets will use the unified password
    - password updates are performed on all wallets
   Whether the password is unified can be seen in the GUI: In the
   'Settings' dialog, the description for the password setting is
   'Change password for this wallet' if the password is not unified,
   and becomes 'Change password' if password is unified.
 * Android PIN reset: If the password is unified, the PIN can be reset
   by providing the password.
 * Android: fees have been removed from the settings dialog. Instead,
   the fee slider is shown everytime an transaction will be performed.
 * Block explorer: allow setting a custom URL in Qt GUI.
 * fix Kivy bug with scrollbar
 * fix localization issues
 * Qt GUI: fix some crashes when exiting, make sure pressing
   Ctrl-C always quits
 * Android: use more modern application launcher/icon,
   implement freezing addresses
 * enhance protx code, add protx info to docs (#237)
 * other minor fixes


# Release 4.0.9.4

 * qt: fix do_export_history (#213)
 * couple of qt style fixes (#216)
 * Fix offline command backup old version on wallet upgrade (#217)
 * add support to save uncompleted multisig wallets (#219)
 * installwizard: show hw xpub after adding cosigner (#220)
 * Add plugin scan_over_gap (#221)
 * qt: add show_utxo_time config/qt option (#222)
 * Split dash_ps module (#223)
 * dash_ps: allow set absolute denoms count to mix (#224)
 * fixes for privatesend with hw wallet (#225)


# Release 4.0.9.3

 * kivy: fix history screen tx_type processing (#207)
 * qt: pass InvoiceExt data to do_pay_invoice (#209)
 * kivy/qt: fix tx_type handling (#210)


# Release 4.0.9.2

 * qt: fix PayToEdit.parse_amount (#172)
 * gui qt/kivy: add ProTx/LLMQ reset UI (#173)
 * PS: separate new denoms/collateral txins by address (#175)
 * qt PS Dialog: fix PS Keystore tab (#176)
 * various DIP3 fixes (post 4.0.9.1) (#177)
 * util.py: fix mainnet Dash.org block explorer (#180)
 * paymentrequest: check network on PaymentRequest parse (#182)
 * qt: fix protected_with_parent decorator (#183)
 * GUI: fix send Payment msg from qt/kivy gui, do not broadcast tx
   when sending Payment to Anypay (#185)
 * invoices: add InvoiceExt with Dash extra data support, which
   includes PrivateSend send status and DIP2 extra data/tx type (#198)


# Release 4.0.9.1

contrib/dash/actions/script-osx.sh
#!/bin/bash
set -ev

export MACOSX_DEPLOYMENT_TARGET=10.15

export PY37BINDIR=/Library/Frameworks/Python.framework/Versions/3.7/bin/
export PATH=$PATH:$PY37BINDIR
echo osx build version is $DASH_ELECTRUM_VERSION


if [[ -n $GITHUB_REF ]]; then
    PIP_CMD="sudo python3 -m pip"
else
    python3 -m virtualenv env
    source env/bin/activate
    PIP_CMD="pip"
fi


$PIP_CMD install --no-dependencies --no-warn-script-location -I \
    -r contrib/deterministic-build/requirements.txt
$PIP_CMD install --no-dependencies --no-warn-script-location -I \
    -r contrib/deterministic-build/requirements-hw.txt
$PIP_CMD install --no-dependencies --no-warn-script-location -I \
    -r contrib/deterministic-build/requirements-binaries-mac.txt
$PIP_CMD install --no-dependencies --no-warn-script-location -I x11_hash>=1.4

$PIP_CMD install --no-dependencies --no-warn-script-location -I \
    -r contrib/deterministic-build/requirements-build-mac.txt

export PATH="/usr/local/opt/gettext/bin:$PATH"
./contrib/make_locale
find . -name '*.po' -delete
find . -name '*.pot' -delete

cp contrib/osx/osx_actions.spec osx.spec
cp contrib/dash/pyi_runtimehook.py .
cp contrib/dash/pyi_tctl_runtimehook.py .

pyinstaller --clean \
    -y \
    --name electrum-dash-$DASH_ELECTRUM_VERSION.bin \
    osx.spec

sudo hdiutil create -fs HFS+ -volname "Dash Electrum" \
    -srcfolder dist/Dash\ Electrum.app \
    dist/Dash-Electrum-$DASH_ELECTRUM_VERSION-macosx.dmg


contrib/dash/actions/install-osx.sh
#!/bin/bash
set -ev

if [[ -n $GITHUB_ACTION ]]; then
    PYTHON_VERSION=3.7.9
    PYFTP=https://www.python.org/ftp/python/$PYTHON_VERSION
    PYPKG_NAME=python-$PYTHON_VERSION-macosx10.9.pkg
    PY_SHA256=bf54a14eef23467991e8c7a88c7307762e484c024a94ec1ee292ac1db3d41fc9
    echo "$PY_SHA256  $PYPKG_NAME" > $PYPKG_NAME.sha256
    curl -O $PYFTP/$PYPKG_NAME
    shasum -a256 -s -c $PYPKG_NAME.sha256
    sudo installer -pkg $PYPKG_NAME -target /
    rm $PYPKG_NAME $PYPKG_NAME.sha256
fi

LIBUSB_VER=1.0.24
LIBUSB_URI=https://github.com/libusb/libusb/releases/download
LIBUSB_SHA=7efd2685f7b327326dcfb85cee426d9b871fd70e22caa15bb68d595ce2a2b12a
LIBUSB_FILE=libusb-${LIBUSB_VER}.tar.bz2
echo "${LIBUSB_SHA}  ${LIBUSB_FILE}" > ${LIBUSB_FILE}.sha256
curl -O -L ${LIBUSB_URI}/v${LIBUSB_VER}/${LIBUSB_FILE}
tar -xzvf ${LIBUSB_FILE}
shasum -a256 -s -c ${LIBUSB_FILE}.sha256
pushd libusb-${LIBUSB_VER}
./configure --disable-dependency-tracking --prefix=/opt/libusb
sudo env MACOSX_DEPLOYMENT_TARGET=10.15 make install
popd
sudo rm -rf libusb-${LIBUSB_VER}*
cp /opt/libusb/lib/libusb-1.*.dylib .

LSECP256K1_PATH=https://github.com/zebra-lucky/secp256k1/
LSECP256K1_PATH=${LSECP256K1_PATH}releases/download/210521
LSECP256K1_FILE=libsecp256k1-210521-osx.tgz
LIB_SHA256=51c861bfb894ec520cc1ee0225fae00447aa86096782a1acd1fc6e338a576ea7
echo "$LIB_SHA256  $LSECP256K1_FILE" > $LSECP256K1_FILE.sha256
curl -O -L ${LSECP256K1_PATH}/${LSECP256K1_FILE}
shasum -a256 -s -c ${LSECP256K1_FILE}.sha256
tar -xzf ${LSECP256K1_FILE}
rm -f libsecp256k1.0.dylib
cp libsecp256k1/libsecp256k1.0.dylib .
rm -rf libsecp256k1/ ${LSECP256K1_FILE} ${LSECP256K1_FILE}.sha256

brew install gettext libtool automake pkg-config

if [[ -n $GITHUB_REF ]]; then
    echo "Building ZBar dylib..."
    rm -f libzbar.0.dylib
    export MACOSX_DEPLOYMENT_TARGET=10.15
    ./contrib/make_zbar.sh
    rm -rf contrib/zbar/
fi


contrib/dash/travis/travis-build-osx.sh
#!/bin/bash
set -ev

export MACOSX_DEPLOYMENT_TARGET=10.15

export PY37BINDIR=/Library/Frameworks/Python.framework/Versions/3.7/bin/
export PATH=$PATH:$PY37BINDIR
source ./contrib/dash/travis/electrum_dash_version_env.sh;
echo osx build version is $DASH_ELECTRUM_VERSION


cd build
if [[ -n $TRAVIS_TAG ]]; then
    BUILD_REPO_URL=https://github.com/ktechmidas/electrum-dash.git
    git clone --branch $TRAVIS_TAG $BUILD_REPO_URL electrum-dash
    PIP_CMD="sudo python3 -m pip"
else
    git clone .. electrum-dash
    python3 -m virtualenv env
    source env/bin/activate
    PIP_CMD="pip"
fi
cd electrum-dash


$PIP_CMD install --no-dependencies --no-warn-script-location -I \
    -r contrib/deterministic-build/requirements.txt
$PIP_CMD install --no-dependencies --no-warn-script-location -I \
    -r contrib/deterministic-build/requirements-hw.txt
$PIP_CMD install --no-dependencies --no-warn-script-location -I \
    -r contrib/deterministic-build/requirements-binaries-mac.txt
$PIP_CMD install --no-dependencies --no-warn-script-location -I x11_hash>=1.4

$PIP_CMD install --no-dependencies --no-warn-script-location -I \
    -r contrib/deterministic-build/requirements-build-mac.txt

export PATH="/usr/local/opt/gettext/bin:$PATH"
./contrib/make_locale
find . -name '*.po' -delete
find . -name '*.pot' -delete

cp contrib/osx/osx.spec .
cp contrib/dash/pyi_runtimehook.py .
cp contrib/dash/pyi_tctl_runtimehook.py .

pyinstaller --clean \
    -y \
    --name electrum-dash-$DASH_ELECTRUM_VERSION.bin \
    osx.spec

sudo hdiutil create -fs HFS+ -volname "Dash Electrum" \
    -srcfolder dist/Dash\ Electrum.app \
    dist/Dash-Electrum-$DASH_ELECTRUM_VERSION-macosx.dmg
